pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -87,38 +114,43 @@ type kubenetNetworkPlugin struct {
 	hostportSyncer  hostport.HostportSyncer
 	hostportManager hostport.HostPortManager
 	iptables        utiliptables.Interface
+	iptablesv6      utiliptables.Interface
 	sysctl          utilsysctl.Interface
 	ebtables        utilebtables.Interface
 	// binDirs is passed by kubelet cni-bin-dir parameter.
 	// kubenet will search for CNI binaries in DefaultCNIDir first, then continue to binDirs.
-	binDirs           []string
-	nonMasqueradeCIDR string
-	podCidr           string
-	gateway           net.IP
+	binDirs            []string
+	nonMasqueradeCIDRs []string
+	podCIDRs           []*net.IPNet
+	podGateways        []net.IP
 }
 
 func NewPlugin(networkPluginDirs []string) network.NetworkPlugin {
-	protocol := utiliptables.ProtocolIpv4
 	execer := utilexec.New()
 	dbus := utildbus.New()
-	sysctl := utilsysctl.New()
-	iptInterface := utiliptables.New(execer, dbus, protocol)
+	iptInterface := utiliptables.New(execer, dbus, utiliptables.ProtocolIpv4)
+	iptInterfacev6 := utiliptables.New(execer, dbus, utiliptables.ProtocolIpv6)
 	return &kubenetNetworkPlugin{
-		podIPs:            make(map[kubecontainer.ContainerID]string),
-		execer:            utilexec.New(),
-		iptables:          iptInterface,
-		sysctl:            sysctl,
-		binDirs:           append([]string{DefaultCNIDir}, networkPluginDirs...),
-		hostportSyncer:    hostport.NewHostportSyncer(iptInterface),
-		hostportManager:   hostport.NewHostportManager(iptInterface),
-		nonMasqueradeCIDR: "10.0.0.0/8",
+		podIPs:          make(map[kubecontainer.ContainerID]utilsets.String),
+		execer:          utilexec.New(),
+		iptables:        iptInterface,
+		iptablesv6:      iptInterfacev6,
+		sysctl:          utilsysctl.New(),
+		binDirs:         append([]string{DefaultCNIDir}, networkPluginDirs...),
+		hostportSyncer:  hostport.NewHostportSyncer(iptInterface),
+		hostportManager: hostport.NewHostportManager(iptInterface),
+		//TODO (khenidak): maybe it should be defaulted to zero address?
+		// 10.0.0.0/8 is the default value before dualstack changes
+		nonMasqueradeCIDRs: []string{"10.0.0.0/8"},
+		podCIDRs:           make([]*net.IPNet, 0),
+		podGateways:        make([]net.IP, 0),
 	}
 }
 
-func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDR string, mtu int) error {
+func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDRs string, mtu int) error {
```

Comment:
Passing around CSV deep inside the code is a bit of a smell. nonMasqueradeCIDRs should be a `[]string`